src/node/txreconciliation.cpp

```diff
@@ -357,79 +357,80 @@ class TxReconciliationTracker::Impl
-        // In order to do this, we sort the peers in random order, and then we pick the top `n` peers of the resulting collection.
-        // If our chosen peer happens to be within the picked selection, we fanout to it, otherwise we reconcile.
-        const size_t targets_size = ((deterministic_randomizer.Finalize() & 0xFFFFFFFF) + uint64_t(n * 0x100000000)) >> 32;
-
-        std::vector<std::pair<uint64_t, NodeId>> best_peers;
-        best_peers.reserve(m_states.size());
+        AssertLockNotHeld(m_txreconciliation_mutex);
+        LOCK(m_txreconciliation_mutex);
+
+        // We decide whether a particular peer is a low-fanout flood target differently based on its connection direction:
+        // - for outbounds we have a fixed number of flood destinations.
+        // - for inbounds we use a fraction of all inbound peers supporting tx relay.
+        size_t outbounds_target_size = 0;
+        if (OUTBOUND_FANOUT_DESTINATIONS > outbounds_fanout_tx_relay) {
+            outbounds_target_size = OUTBOUND_FANOUT_DESTINATIONS - outbounds_fanout_tx_relay;
+        }
 
+        // Since we use the fraction for inbound peers, we first need to compute the total number of inbound targets.
+        const double inbound_targets = (inbounds_fanout_tx_relay + m_inbounds_count) * INBOUND_FANOUT_DESTINATIONS_FRACTION;
+        double n = std::max(inbound_targets - inbounds_fanout_tx_relay, 0.0);
+
+        // Being this a fraction, we need to round it either up or down. We do this deterministically at random based on the
+        // transaction we are picking the peers for.
+        CSipHasher deterministic_randomizer_in{m_deterministic_randomizer};
+        deterministic_randomizer_in.Write(wtxid.ToUint256());
+        CSipHasher deterministic_randomizer_out{deterministic_randomizer_in};
+        const size_t inbounds_target_size = ((deterministic_randomizer_in.Finalize() & 0xFFFFFFFF) + uint64_t(n * 0x100000000)) >> 32;
+
+        // Pick all reconciliation registered peers and assign them a deterministically random value based on their peer id
+        // Also, split peers in inbounds/outbounds
+        std::vector<std::pair<uint64_t, NodeId>> sorted_inbounds, sorted_outbounds;
+        sorted_inbounds.reserve(m_inbounds_count);
+        Assume(m_states.size() >= m_inbounds_count);
+        sorted_outbounds.reserve(m_states.size() - m_inbounds_count);
         for (const auto& [node_id, op_peer_state]: m_states) {
             const auto peer_state = std::get_if<TxReconciliationState>(&op_peer_state);
-            if (peer_state && peer_state->m_we_initiate == we_initiate) {
-                uint64_t hash_key = CSipHasher(deterministic_randomizer).Write(node_id).Finalize();
-                best_peers.emplace_back(hash_key, node_id);
+            if (peer_state) {
+                if (peer_state->m_we_initiate) {
+                    uint64_t hash_key = CSipHasher(deterministic_randomizer_out).Write(node_id).Finalize();
```

Comment:
```
        auto assign_key = [](const TxReconciliationState& peer_state, NodeId node_id, CSipHasher randomizer, std::vector<std::pair<uint64_t, NodeId>>& weighted_peers) {
            uint64_t hash_key = randomizer.Write(node_id).Finalize();
            weighted_peers.emplace_back(hash_key, node_id);
        };

        for (const auto& [node_id, op_peer_state]: m_states) {
            const auto peer_state = std::get_if<TxReconciliationState>(&op_peer_state);
            if (peer_state) {
                if (peer_state->m_we_initiate) {
                    assign_key(*peer_state, node_id, deterministic_randomizer_out, sorted_outbounds);
                } else {
                    assign_key(*peer_state, node_id, deterministic_randomizer_in, sorted_inbounds);
                }
            }
        }
```

very subjectively a little cleaner :)

Reply:
I don't think it makes a huge difference, but I'll take it